### DISCUSSION TITLE
Websocket support (Socket.io, etc?)

### DISCUSSION THREAD

**Discussion by micahlt** [vercel/community#422] [Idea]
It seems like Vercel with its supposedly easy-to-set-up serverless functions lacks the ability to use websockets.  This drastically weakens Vercel, and I'm very disappointed.  Is websocket support on the plan for the future?

**Comment by styfle** [comment]
Hi @micahlt 

Serverless Functions do not support WebSockets. ([docs](https://vercel.com/docs/platform/limits#websockets))

Instead of pushing data, you can intelligently fetch data on-demand. For example, the Vercel dashboard delivers real-time updates using [SWR](https://swr.vercel.app).

If you need persistent connections to push data to the client, consider using third party pub/sub services. See the [Pusher Channels Guide](https://vercel.com/guides/deploying-pusher-channels-with-vercel) for an example.

**Reply by hacknlove** [reply]
A bit disapointing.

Maybe some users will be opened to pay some extra, to have this feature.

**Reply by micahlt** [reply]
@hacknlove Indeed - Websockets are a critical part of many web-based applications, and it can't be that hard to support them.  I'd love to see this in the future.

**Reply by paulogdm** [reply]
Pusher.com is another alternative. If you want, you can also host the backend of the app in an external provider.

**Reply by YarivGilad** [reply]
the serverless framework does support WebSockets for a long time now... 
Vercel has other priorities. it is an old request they chose not to go for, this is why they direct you to services like pusher.
We need to accept it and make our own choices.

**Reply by oddnugget** [reply]
@styfle  Just to make sure I understood this correctly, it is only the hosting of a websocket api in a serverless function that isn't possible, right? Connecting to an external websocket api from a page is possible on vercel with next.js?

**Reply by paulogdm** [reply]
Correct!

**Reply by christian-ariel** [reply]
> Hi @micahlt
> 
> Serverless Functions do not support WebSockets. ([docs](https://vercel.com/docs/platform/limits#websockets))
> 
> Instead of pushing data, you can intelligently fetch data on-demand. For example, the Vercel dashboard delivers real-time updates using [SWR](https://swr.vercel.app).
> 
> If you need persistent connections to push data to the client, consider using third party pub/sub services. See the [Pusher Channels Guide](https://vercel.com/guides/deploying-pusher-channels-with-vercel) for an example.

I am trying to implement Pusher with next js and vercel but I can not make it work. Is there any updated documentation on this? I am in touch with the Pusher support but they can not figure it out yet. The events only trigger when refreshing the browser or when switching tabs.

**Comment by JollyGrin** [comment]
Aww man, just discovering this now. I used to have a websocket app live on zeit now but after the switch to vercel it no longer works. Still not sure if this is on my end, but this is a huge bummer if so.

**Reply by 0m15** [reply]
I'm getting `WebSocket connection to 'wss://***.vercel.app/' failed: Error during WebSocket handshake: Unexpected response code: 200` but the exact same server code works both locally and on gae.

Using express + ws + http.

**Reply by micahlt** [reply]
@JollyGrin @0m15 I've now had several maintainers and Vercel support members confirm that Vercel is not compatible with WebSockets, and it's not planned to be added in the future.  It's disappointing, but that's the choice the company has made.

**Comment by jkarneges** [comment]
For folks needing WebSockets: https://blog.fanout.io/2020/09/21/vercel-and-websockets/

**Reply by micahlt** [reply]
Thank you!  It does look like their free plan is quite limited compared to [Pusher](https://pusher.com), however.

**Reply by jkarneges** [reply]
Hi Micah! I'm the Fanout founder. Thanks for the feedback.

Indeed our free plan has smaller limits, but the service is more powerful in some ways. You get real WebSockets and can use a custom domain. If the limits are a problem, I'm happy to discuss it separately (my email is justin at fanout.io).

**Comment by benmccann** [comment]
I don't understand the responses here saying it's not possible - perhaps it's just because this is an old thread. However, today, [Vercel Edge Functions are built on Cloudflare Workers](https://news.ycombinator.com/item?id=29003514), which [do appear to support websockets](https://developers.cloudflare.com/workers/runtime-apis/websockets/use-websockets/).

**Reply by drattansingh** [reply]
Have you tried it to confirm that it's working in vercel?
